File: csharp/need-for-speed/NeedForSpeed.cs
class RemoteControlCar
{
    private int _battery;

    private int _distance;
    public int BatteryDrain { get; init; }
    public int Speed { get; init; }

    public RemoteControlCar(int speed, int batteryDrain)
    {
        _battery = 100;
        BatteryDrain = batteryDrain;
        Speed = speed;
    }

    public int Battery() => _battery;

    public bool BatteryDrained() => _battery < BatteryDrain;

    public int DistanceDriven() => _distance;

    public void Drive()
    {
        if (BatteryDrained()) return;
        _distance += Speed;
        _battery -= BatteryDrain;
    }

    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}

class RaceTrack
{
    private int _distance;

    public RaceTrack(int distance) => _distance = distance;

    public bool TryFinishTrack(RemoteControlCar car)
    {
        if (car.BatteryDrained()) return false;

        int maxDistanceLeft = car.Battery() / car.BatteryDrain * car.Speed;
        return maxDistanceLeft >= _distance;
    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int battery;
    private int batteryDrain;
    private int driven;
    private int speed;

    NeedForSpeed(int speed, int batteryDrain) {
        this.battery = 100;
        this.batteryDrain = batteryDrain;
        this.driven = 0;
        this.speed = speed;
    }

    public boolean batteryDrained() {
        return battery < batteryDrain;
    }

    public int distanceDriven() {
        return driven;
    }

    public int distanceLeft() {
        return (battery / batteryDrain) * speed;
    }

    public void drive() {
        if (battery >= batteryDrain) {
            battery -= batteryDrain;
            driven += speed;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return !car.batteryDrained() && car.distanceLeft() >= distance;
    }
}
